File: 019. Remove Nth Node From End of List/19. Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        fp = sp = head
        for i in xrange(n):
            fp = fp.next
        if not fp:
            return head.next
        while fp.next:
            fp, sp = fp.next, sp.next
        sp.next = sp.next.next
        return head

File: 019. Remove Nth Node From End of List/19. Remove Nth Node From End of List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    struct ListNode *fp = head, *sp = head;
    for (int i = 0; i < n; i++)
        fp = fp -> next;
    if (fp == NULL)
        return head -> next;
    while (fp -> next != NULL){
        fp = fp -> next;
        sp = sp -> next;
    }
    sp -> next = sp -> next -> next;
    return head;
}
